[#371] Add functionality for new keyboard expanded layout

damien-rivet opened the pull request:
### Contributor checklist

- [x] This pull request is on a [separate branch](https://docs.github.com/en/get-started/quickstart/github-flow) and not the main branch

---

### Description

This PR implements #371 by adding functionality for the new expanded layout that was introduced by #352.

The `indent` key is now fully functional as well as the `capslock` key.

#### Recordings

**Old design**

https://github.com/scribe-org/Scribe-iOS/assets/1595177/6a6765a9-b749-4554-a032-5fcba6e322dd

**New design**

https://github.com/scribe-org/Scribe-iOS/assets/1595177/4a0e4c1b-f187-4450-9976-7ad333543457

#### Screenshots

| Description | Screenshot |
|--------|--------|
| **New design (iPad)** | ![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-10-22 at 14 46 46](https://github.com/scribe-org/Scribe-iOS/assets/1595177/15fd622a-9106-4e93-bcf8-11465db63df3) |
| **Design on iPhone (no changes)** | ![Simulator Screenshot - iPhone 15 Pro - 2023-10-22 at 14 54 17](https://github.com/scribe-org/Scribe-iOS/assets/1595177/fe26bd51-a7fb-4251-9293-b4809c80daa6) |
| **Capslock** | ![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-10-22 at 14 46 53](https://github.com/scribe-org/Scribe-iOS/assets/1595177/f5ed8f09-d9c6-4ecb-bd1d-f34fe2f2ba74) |
| **Uppercased suggestion** | ![Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2023-10-22 at 14 56 09](https://github.com/scribe-org/Scribe-iOS/assets/1595177/8d9ceb40-de43-48e8-bb78-e2d0f0db5836) |

### Related issue

- #352

henrikth93 commented:
Tested it and it looks good to me.

andrewtavis commented:
Really appreciate the overall care/detail here and especially some a classic sentence for the example, @damien-rivet! 🦊😊 Going through it all now. Sorry that this took a bit. I needed to switch contexts for another project to do UI/UX design, and that always tends to slow down other work (although fun 🙃).

andrewtavis commented on Keyboards/KeyboardsBase/KeyAnimation.swift:
```diff
@@ -319,7 +319,7 @@ func setPhoneKeyPopCharSize(char: String) {
       keyHoldPopChar.font = .systemFont(ofSize: letterKeyWidth / 1.15)
     }
-  } else if shiftButtonState == .shift || shiftButtonState == .caps {
+  } else if shiftButtonState == .shift || capsLockButtonState == .locked {
```
Very intuitive :)

andrewtavis commented on Keyboards/KeyboardsBase/KeyboardKeys.swift:
```diff
@@ -256,22 +256,14 @@ class KeyboardKey: UIButton {
-      || key == "return"
-      || key == "hideKeyboard"
-    {
+    } else if ["123", ".?123", "return", "hideKeyboard"].contains(key) {
```
Appreciate you fixing these formatting errors! 🙏

andrewtavis commented on Keyboards/KeyboardsBase/KeyboardKeys.swift:
```diff
@@ -291,25 +283,46 @@ class KeyboardKey: UIButton {
     }
+  }
 
+  /// Adjusts the style of the button based on different states
```
Minor nit that I'm noting here is that we'd like this to be a complete sentence with a period :) I can get it though!

andrewtavis commented on Keyboards/KeyboardsBase/KeyboardStyling.swift:
```diff
@@ -40,7 +40,9 @@ var keysThatAreSlightlyLarger = [
   "chevron.right",
   "shift",
   "shift.fill",
+  "capslock",
```
Again the dedication here is amazing. Looking into and finding `keysThatAreSlightlyLarger` is great, but also has me worried about the state of some of the code that we're using so many random variables everywhere.

A basic ask would be if you had ideas for how we can improve the code base, it'd be great if you could open some issues. Anything that you've seen that might be some solid refactoring would be very welcome!

damien-rivet replied:
Good point 🤔 I'll do a quick sweep over the codebase and open a few issues for code quality ^^

andrewtavis replied:
Thanks so much, @damien-rivet! Always looking for some `good first issues` for new folks to work on 😊

andrewtavis commented on Keyboards/KeyboardsBase/KeyboardViewController.swift:
```diff
@@ -406,7 +406,9 @@ class KeyboardViewController: UIInputViewController {
           if completionOptions.count <= 3 {
             while i < completionOptions.count {
-              if shiftButtonState == .caps {
+              if shiftButtonState == .shift {
```
Interesting, we were missing shift?? Great catch :)

damien-rivet replied:
Yup, suggestions were not handled properly for capitalised first-letter but it will be history after the merge.

andrewtavis commented on Keyboards/KeyboardsBase/KeyboardViewController.swift:
```diff
@@ -1925,6 +1937,10 @@ class KeyboardViewController: UIInputViewController {
           // Set the width of the key given device and the given key.
           btn.adjustKeyWidth()
+
+          // Update the button style
```
Noting the comment for myself.

andrewtavis commented on Keyboards/KeyboardsBase/KeyboardViewController.swift:
```diff
@@ -2482,7 +2498,14 @@ class KeyboardViewController: UIInputViewController {
     case "shift":
-      shiftButtonState = shiftButtonState == .normal ? .shift : .normal
+      if capsLockButtonState == .locked {
+        // Return capitalization to default
```
Nice as well as we now have both states to account for :)

andrewtavis commented on Keyboards/KeyboardsBase/KeyboardViewController.swift:
```diff
@@ -2657,6 +2684,15 @@ class KeyboardViewController: UIInputViewController {
     }
   }
 
+  private func switchToFullCaps() {
```
Another thing that we've been dealing with is that not enough of the code base is split into functions. I do some rounds from time to time and break some things out, but if you came across some parts where we can extract some code then by all means let me know or open an issue!

andrewtavis commented on Keyboards/KeyboardsBase/KeyboardKeys.swift:
```diff
@@ -291,25 +283,46 @@ class KeyboardKey: UIButton {
+      if [.translate, .conjugate, .plural].contains(commandState) {
+        // Color the return key depending on if it's being used as enter for commands.
+        backgroundColor = commandKeyColor
+      }
```
We need the else back here for one bug I'm about to mention such that we get `backgroundColor = specialKeyColor`.

damien-rivet replied:
There's a default case which should handle the `else`but I'll do a review of my code to see if it's not missing some cases.

andrewtavis commented on Keyboards/KeyboardsBase/KeyboardKeys.swift:
```diff
@@ -291,25 +283,46 @@ class KeyboardKey: UIButton {
         styleIconBtn(btn: self, color: UIColor.label, iconName: "shift.fill")
-      } else if shiftButtonState == .caps {
-        backgroundColor = keyPressedColor
-        styleIconBtn(btn: self, color: UIColor.label, iconName: "capslock.fill")
```
Losing this line is disabling an indication that we're in caps lock on iPhones :)

andrewtavis commented:
Two things I'm seeing based on testing:

- For iPhones we now have that the shift key doesn't change to the caps lock key when it's double tapped
  - The caps lock functionality still works, but there's no change in the button to indicate that it will
  - See [this comment](https://github.com/scribe-org/Scribe-iOS/pull/373#discussion_r1375560952)
- The return key is changing to Scribe blue when we're in command mode, but it's white instead of grey for a special key when we're in the normal keyboard functionality (see screens)
  - See [this comment](https://github.com/scribe-org/Scribe-iOS/pull/373#discussion_r1375560671)

Aside from this, really amazing work and one of our best first time contributions, @damien-rivet! I'm happy to get to the final fixes myself. Thank you!

damien-rivet commented:
@andrewtavis I've fixed both issues in my last commit.

While testing I discovered a crash (nil unwrapping) in Command mode, I've added some checks to prevent those from happening again.

andrewtavis commented:
Fantastic, @damien-rivet! I'm off to dinner, but will get to the final review this evening :) :)

andrewtavis reviewed (approved):
Thanks for getting to the changes so quickly, @damien-rivet, as well as finding and fixing the nil unwrapping! Really great to have this contribution :) Will bring it in and check to make sure that we have everything updated as far as the changelog and such, and then will close out the issue 😊
